### PR TITLE
feat: support wfr cancel

### PR DESCRIPTION
**Pull request by cd1989**
**What this PR does / why we need it**:

Support WorkflowRun cancel

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @supereagle @zhujian7 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/workflow/workflowrun/operator.go [hunk 1]
```diff
@@ -480,3 +482,31 @@ func (o *operator) GC(lastTry, wfrDeletion bool) error {
...
+
+// Cancel cancels execution of the WorkflowRun, it will remove all pods of it.
+func (o *operator) Cancel() {
+	for stg, status := range o.wfr.Status.Stages {
+		if status.Pod == nil {
+			continue
+		}
+
+		// If the stage already in terminated status, skip it.
+		if status.Status.Phase == v1alpha1.StatusSucceeded ||
```
Comment (supereagle) [inline pkg/workflow/workflowrun/operator.go]: This logic is frequently used, define a method for it?

File: pkg/workflow/workflowrun/operator.go [hunk 2]
```diff
@@ -399,8 +410,9 @@ func (o *operator) GC(lastTry, wfrDeletion bool) error {
 			if !wfrDeletion {
 				o.recorder.Eventf(o.wfr, corev1.EventTypeWarning, "GC", "Delete pod '%s' error: %v", status.Pod.Name, err)
 			}
+		} else {
+			log.WithField("ns", status.Pod.Namespace).WithField("pod", status.Pod.Name).Info("Pod deleted")
```
Comment (zhujian7) [inline pkg/workflow/workflowrun/operator.go]: Why do this change?

File: pkg/workflow/controller/handlers/workflowrun/handler.go [hunk 3]
```diff
@@ -107,7 +107,7 @@ func (h *Handler) ObjectUpdated(obj interface{}) {
 	// the GC queue.
 	h.GCProcessor.Add(originWfr)
 
-	// If the WorkflowRun has already been terminated(Completed, Error, Cancel), send notifications if necessary,
+	// If the WorkflowRun has already been terminated(Completed, Error), send notifications if necessary,
```
Comment (zhujian7) [inline pkg/workflow/controller/handlers/workflowrun/handler.go]: I think we also need to send notify even if it is canceled, @supereagle  WDYT?

File: pkg/workflow/workflowrun/utils.go [hunk 4]
```diff
@@ -16,13 +16,13 @@ import (
 func resolveStatus(latest, update *v1alpha1.Status) *v1alpha1.Status {
 	// If the latest status is already a terminated status (Completed, Error), no need to update
```
Comment (zhujian7) [inline pkg/workflow/workflowrun/utils.go]: terminated status should be `Succeeded, Failed and Cancelled`

File: pkg/workflow/controller/handlers/workflowrun/handler.go (hunk 3 above)
Comment (supereagle) [reply pkg/workflow/controller/handlers/workflowrun/handler.go]: In my next PR, will send notification for all wfrs.

**Pull request merged**